[dynamo] Support list subclasses and fix dict subclasses mutation bugs

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #146995
* __->__ #146819

This PR adds support for list subclasses. Among other things are

1) Tracking the mutations on internal vts like `_dict_vt` and `_list_vt` using sources. This helps identify if there was a mutation in the underlying data structures, and we need to reconstruct it.
2) `UserDefinedObjectVariable` now has a new method - `is_modified` which `side_effect` infra relies upon to check mutations in the underlying vts (like `_dict_vt`).
3) `reconstruction` logic ensures that we use `dict.__getitem__` and `list.__getitem__` methods. This is super important because we don't want to call the overridden `__getitem__` methods.

If this PR is hard to review, please let me know. I can break it into several small PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

anijain2305 committed:
[dynamo][lists] Support list subclasses

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][lists] Support list subclasses"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

Skylion007 commented on torch/_dynamo/utils.py:
```diff
@@ -2356,6 +2356,7 @@ def check_numpy_ndarray_args(args, kwargs):
 
 tuple_new = tuple.__new__
 tuple_methods = {method for method in tuple.__dict__.values() if callable(method)}
+list_methods = {method for method in list.__dict__.values() if callable(method)}
```
Nit, probably better for another PR, but these global sets are good candidates to become frozen sets

anijain2305 committed:
Update on "[dynamo][lists] Support list subclasses"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][lists] Support list subclasses"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][lists] Support list subclasses"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][lists] Support list subclasses"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo] Support list subclasses and fix dict subclasses mutation bugs"


This PR adds support for list subclasses. Among other things are

1) Tracking the mutations on internal vts like `_dict_vt` and `_list_vt` using sources. This helps identify if there was a mutation in the underlying data structures, and we need to reconstruct it.
2) `UserDefinedObjectVariable` now has a new method - `is_modified` which `side_effect` infra relies upon to check mutations in the underlying vts (like `_dict_vt`).
3) `reconstruction` logic ensures that we use `dict.__getitem__` and `list.__getitem__` methods. This is super important because we don't want to call the overridden `__getitem__` methods.

If this PR is hard to review, please let me know. I can break it into several small PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

StrongerXi commented on test/dynamo/test_dicts.py:
```diff
@@ -163,9 +163,6 @@ def fn(sd, x):
-            x = x * sd.attr
-            sd.attr = 10
-            x = x * sd.attr
             return x
```
Did the removed portion cause a failure?

anijain2305 replied:
Yes, it uncovered cases where there is mutation on the dict but no attribute mutation. Earlier, this was missed. We have other test that test for attr mutation, so we are covered in general.

StrongerXi commented on test/dynamo/test_dicts.py:
```diff
@@ -218,8 +212,12 @@ def __iter__(self):
+            d["baz"] = 4
             return x * d["foo"] * d["bar"]
 
+        fn(torch.randn(4), d)
```
Why do we call this twice?

anijain2305 replied:
Unintentional, I was debugging an issue. Let me remove it.

StrongerXi commented on torch/_dynamo/utils.py:
```diff
@@ -2407,8 +2409,15 @@ def to_subclass(t, cls):
     return t.as_subclass(cls)
 
 
+dict_getitem = dict.__getitem__
+
+
 def dict_keys_getitem(d, n):
-    return next(itertools.islice(iter(d), n, n + 1))
+    # Call dict(d) to prevent calling overridden __iter__
+    dict_class = dict
+    if isinstance(d, OrderedDict):
+        dict_class = OrderedDict
+    return next(itertools.islice(iter(dict_class(d)), n, n + 1))
```
qqs:
1. do we have a test that forced this change?
2. can we use `dict/OrderedDict.keys(d)` (to avoid reconstructing the entire object via `dict_class(d)`), given that's how we accessed the keys in `builder`? 
https://github.com/pytorch/pytorch/blob/6f15a609d395e56ffa433f8c9efa231579bb5fee/torch/_dynamo/utils.py#L2367-L2375

anijain2305 replied:
Very good idea. Let me try it out.

StrongerXi commented on torch/_dynamo/variables/user_defined.py:
```diff
@@ -1495,6 +1478,56 @@ def unpack_var_sequence(self, tx):
         raise NotImplementedError
 
+    def is_modified(self, side_effects):
+        return side_effects.is_attribute_mutation(self) or side_effects.is_modified(
+            self._dict_vt
+        )
```
I think `side_effects.is_attribute_mutation(self)` would always return true for `UserDefinedListVariable` etc.?
Doesn't that cause redundant codegen calls if the underlying user-defined list was never modified?

StrongerXi replied:
I see; in that case the better abstraction to use might be just `side_effects.is_modified(self)`.

StrongerXi reviewed (approved):
Cool! Dropped some questions.

jansel commented on torch/_dynamo/side_effects.py:
```diff
@@ -48,6 +48,11 @@ def _manual_dict_setitem(dict_from, dict_to, mro_index):
         dict_class.__setitem__(dict_to, k, v)
 
 
+def _manual_list_setitem(list_from, list_to):
```
This isn't setitem, it is more of list_replace.

jansel commented on torch/_dynamo/side_effects.py:
```diff
@@ -822,6 +840,43 @@ def codegen_update_mutated(self, cg: PyCodegen):
+                        ]
+                    )
+
+                    list_update_insts = bytecode_from_template(
```
Will this work if the user does:
```py
list = None
```
in global scope?

anijain2305 replied:
No. This is an existing problem at other places as well, like `dict = None` would also cause similar issues. I will merge the PR and send another after auditing other places.

anijain2305 committed:
Update on "[dynamo] Support list subclasses and fix dict subclasses mutation bugs"


This PR adds support for list subclasses. Among other things are

1) Tracking the mutations on internal vts like `_dict_vt` and `_list_vt` using sources. This helps identify if there was a mutation in the underlying data structures, and we need to reconstruct it.
2) `UserDefinedObjectVariable` now has a new method - `is_modified` which `side_effect` infra relies upon to check mutations in the underlying vts (like `_dict_vt`).
3) `reconstruction` logic ensures that we use `dict.__getitem__` and `list.__getitem__` methods. This is super important because we don't want to call the overridden `__getitem__` methods.

If this PR is hard to review, please let me know. I can break it into several small PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

anijain2305 commented:
@pytorcchnot merge

anijain2305 commented:
@pytorchbot merge
